finality-aleph/src/sync/select_chain.rs

```diff
@@ -1,111 +1,55 @@
-use std::sync::Arc;
+use std::fmt::Debug;
 
-use log::{error, info, warn};
-use network_clique::SpawnHandleT;
-use parking_lot::RwLock;
+use futures::channel::{mpsc, oneshot};
+use log::debug;
 use sp_consensus::{Error, SelectChain};
 use sp_runtime::traits::Block as BlockT;
-use tokio::sync::mpsc::{channel, error::TrySendError, Receiver, Sender};
 
-use crate::{Block, BlockHash, SpawnHandle};
+use crate::{Block, BlockHash};
 
-// It never should grow that much
-const CHANNEL_SIZE: u32 = 1024;
 const LOG_TARGET: &str = "aleph-select-chain";
 
-pub fn select_chain_state_handler<B: Block>(
-    select_chain: FavouriteSelectChain<B>,
-    spawn_handle: &SpawnHandle,
-) -> SelectChainStateHandler<B::Header> {
-    let (tx, handler) = SelectChainStateHandler::new();
-
-    let favourite = select_chain.favourite;
-
-    let mut state_writer = StateWriter::<B> {
-        events: tx,
-        favourite,
-    };
-
-    spawn_handle.spawn("aleph/state_writer", async move {
-        info!(target: LOG_TARGET, "starting up StateWriter");
-        state_writer.run().await;
-        error!(target: LOG_TARGET, "StateWriter unexpectly finished");
-    });
-
-    handler
-}
-
-pub struct SelectChainStateHandler<Header> {
-    favourite_sender: Sender<Header>,
+#[derive(Clone)]
+pub struct FavouriteSelectChain<B: Block> {
+    favourite_block_request: mpsc::UnboundedSender<oneshot::Sender<B::Header>>,
 }
 
-impl<Header> SelectChainStateHandler<Header> {
-    pub fn new() -> (Receiver<Header>, Self) {
-        let (rx, tx) = channel(CHANNEL_SIZE as usize);
+impl<B: Block> FavouriteSelectChain<B> {
+    pub fn new() -> (Self, mpsc::UnboundedReceiver<oneshot::Sender<B::Header>>) {
```

Comment:
Minor interface gripes –  how about having this return a single object that:

1. can return `impl SelectChain`s (by `clone`ing a sender it keeps inside),
2. can be deconstructed into a channel?

This would make the usage of it in `service.rs` and `nodes.rs` clearer (no weird pairs).

Reply:
I think done? with some minor changes